src/batch/benches/hash_join.rs

```diff
@@ -171,60 +173,20 @@ fn create_hash_join_executor(
     ))
 }
 
-async fn execute_hash_join_executor(executor: BoxedExecutor) {
-    let mut stream = executor.execute();
-    while let Some(ret) = stream.next().await {
-        black_box(ret.unwrap());
-    }
-}
-
 fn bench_hash_join(c: &mut Criterion) {
-    const LEFT_SIZE: usize = 2 * 1024;
-    const RIGHT_SIZE: usize = 2 * 1024;
-    let rt = Runtime::new().unwrap();
-    for with_cond in [false, true] {
-        for join_type in &[
-            JoinType::Inner,
-            JoinType::LeftOuter,
-            JoinType::LeftSemi,
-            JoinType::LeftAnti,
-            JoinType::RightOuter,
-            JoinType::RightSemi,
-            JoinType::RightAnti,
-        ] {
-            for chunk_size in &[32, 128, 512, 1024] {
-                c.bench_with_input(
-                    BenchmarkId::new(
-                        "HashJoinExecutor",
-                        format!(
-                            "{}({:?})(non_equi_join: {})",
-                            chunk_size, join_type, with_cond
-                        ),
-                    ),
-                    chunk_size,
-                    |b, &chunk_size| {
-                        let left_chunk_num = LEFT_SIZE / chunk_size;
-                        let right_chunk_num = RIGHT_SIZE / chunk_size;
-                        b.to_async(&rt).iter_batched(
-                            || {
-                                create_hash_join_executor(
-                                    *join_type,
-                                    with_cond,
-                                    chunk_size,
-                                    left_chunk_num,
-                                    chunk_size,
-                                    right_chunk_num,
-                                )
-                            },
-                            |e| execute_hash_join_executor(e),
-                            BatchSize::SmallInput,
-                        );
-                    },
-                );
-            }
-        }
-    }
+    let with_conds = vec![false, true];
+    let join_types = vec![
+        JoinType::Inner,
+        JoinType::LeftOuter,
+        JoinType::LeftSemi,
+        JoinType::LeftAnti,
+        JoinType::RightOuter,
+        JoinType::RightSemi,
+        JoinType::RightAnti,
+    ];
+    bench_hash_join_internal(c, with_conds, join_types);
 }
 
+bench_join!("HashJoin");
```

Comment:
I don't think it's a great idea to reuse code using macro in benches, cuz benches are something like unit tests, being clear/WYSIWYG is important. But I'm not discouraging you from doing this, just some humble opinion. It's up to you!☺️

Reply:
There are lots of duplicated code in benching join, so I want to try my best to reduce the duplicates.

Reply:
I agree with @richardchien and we should use type systems rather than macros to avoid duplication. Macros ar e difficult to understand and maintain to me.